[WGSL] Add command line tool to run the compiler

tadeuzagallo opened the pull request:
#### c1fb4e9f678eb1aacf9b056070aced47d70502f7
<pre>
[WGSL] Add command line tool to run the compiler
<a href="https://bugs.webkit.org/show_bug.cgi?id=253892">https://bugs.webkit.org/show_bug.cgi?id=253892</a>
rdar://106708690

Reviewed by Mike Wyrzykowski.

Add a `wgslc` command line executable to invoke the compiler. This will be used to
add compiler tests without requiring the rest of WebGPU.

* Source/WebGPU/Configurations/wgslc.xcconfig: Added.
* Source/WebGPU/WGSL/wgslc.cpp: Added.
(printUsageStatement):
(CommandLine::CommandLine):
(CommandLine::file const):
(CommandLine::entrypoint const):
(CommandLine::dumpASTAfterCheck const):
(CommandLine::dumpASTAtEnd const):
(CommandLine::dumpGeneratedCode const):
(CommandLine::parseArguments):
(runWGSL):
(main):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU.xcodeproj/xcshareddata/xcschemes/WebGPU.xcscheme:

Canonical link: <a href="https://commits.webkit.org/261911@main">https://commits.webkit.org/261911@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0273cd13b49a741961a043fdf421802e32e5bcb5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/113126 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22258 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1796 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4897 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121589 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23629 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13444 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/6143 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118913 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17558 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100817 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106208 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99512 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1355 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46583 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14561 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1397 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/98836 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15276 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10733 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20569 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53387 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8325 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/17122 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/wgslc.cpp:
```diff
@@ -0,0 +1,144 @@
+
+static NO_RETURN void printUsageStatement(bool help = false)
+{
+    fprintf(stderr, "Usage: wgsl [options] <file> <entrypoint>\n");
```
I don't understand. The input to the compiler is way more than a single file and entrypoint. And the output is way more than just a single source string.

litherum replied:
I also don't really understand how this is going to work with the 2-phase compilation design that WGSL has

tadeuzagallo replied:
I'm likely missing something, but my thinking was that the input can be more, and the output can be more, but does it *need* to be more? Here I'm assuming that we'll use the 'auto' layout and will only output the string. This is mostly useful for validating type checking error messages, inferred types and AST transformations, it's not a full standalone compiler. Does that make sense? Otherwise, what do you think the input should be?

mwyrzykowski replied:
I kind of viewed this as a good first step for testing / validation purposes. Maybe we can expand upon it as needed but initially it is sufficient if it is useful for @tadeuzagallo?

litherum replied:
I guess it's fine, and I shouldn't let the perfect be the enemy of the good. As Mike says, we can expand up on it as necessary later.

litherum commented on Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
```diff
@@ -986,6 +1085,51 @@
+			isa = XCBuildConfiguration;
+			buildSettings = {
+				CODE_SIGN_STYLE = Automatic;
+				PRODUCT_NAME = "$(TARGET_NAME)";
```
Please delete these

tadeuzagallo replied:
What are "these", the two entries under `buildSettings`? This was generated by Xcode when I created the target, and I'm not entirely sure what to keep or remove.

litherum replied:
Yes, those 2 entries. Any build configurations should be in .xcconfig files, not here.

litherum commented on Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
```diff
@@ -986,6 +1085,51 @@
+			isa = XCBuildConfiguration;
+			buildSettings = {
+				CODE_SIGN_STYLE = Automatic;
+				PRODUCT_NAME = "$(TARGET_NAME)";
```
Ditto

litherum commented on Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
```diff
@@ -986,6 +1085,51 @@
+			isa = XCBuildConfiguration;
+			buildSettings = {
+				CODE_SIGN_STYLE = Automatic;
+				PRODUCT_NAME = "$(TARGET_NAME)";
```
Ditto

litherum commented on Source/WebGPU/WebGPU.xcodeproj/xcshareddata/xcschemes/WebGPU.xcscheme:
```diff
@@ -20,6 +20,20 @@
                ReferencedContainer = "container:WebGPU.xcodeproj">
             </BuildableReference>
          </BuildActionEntry>
+         <BuildActionEntry
+            buildForTesting = "YES"
+            buildForRunning = "YES"
+            buildForProfiling = "YES"
+            buildForArchiving = "YES"
+            buildForAnalyzing = "YES">
+            <BuildableReference
+               BuildableIdentifier = "primary"
+               BlueprintIdentifier = "97FA1A8F29C086CE0052D650"
+               BuildableName = "All"
+               BlueprintName = "All"
+               ReferencedContainer = "container:WebGPU.xcodeproj">
+            </BuildableReference>
+         </BuildActionEntry>
```
I don't think this is supposed to be part of the commit

tadeuzagallo replied:
Why? I created an `All` target to match what we do in other projects, but I can remove it.

litherum replied:
oh ok, keeping sounds fine then

litherum commented on Source/WebGPU/WGSL/wgslc.cpp:
```diff
@@ -0,0 +1,144 @@
+    if (options.dumpASTAfterCheck())
+        WGSL::AST::dumpAST(shaderModule);
+
+    String entrypointName = String::fromLatin1(options.entrypoint());
```
Are you sure latin1 is right for this?

tadeuzagallo replied:
I didn't anticipate passing any other kinds of strings in test programs, but maybe it's useful? Are you implying it should be utf8?

litherum replied:
yeah, utf8 seems better 🤔

tadeuzagallo committed:
[WGSL] Add command line tool to run the compiler
https://bugs.webkit.org/show_bug.cgi?id=253892
rdar://106708690

Reviewed by Mike Wyrzykowski.

Add a `wgslc` command line executable to invoke the compiler. This will be used to
add compiler tests without requiring the rest of WebGPU.

* Source/WebGPU/Configurations/wgslc.xcconfig: Added.
* Source/WebGPU/WGSL/wgslc.cpp: Added.
(printUsageStatement):
(CommandLine::CommandLine):
(CommandLine::file const):
(CommandLine::entrypoint const):
(CommandLine::dumpASTAfterCheck const):
(CommandLine::dumpASTAtEnd const):
(CommandLine::dumpGeneratedCode const):
(CommandLine::parseArguments):
(runWGSL):
(main):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU.xcodeproj/xcshareddata/xcschemes/WebGPU.xcscheme:

Canonical link: https://commits.webkit.org/261911@main
